### PR TITLE
Feature/add pipeline id and plugin name to logs

### PR DESCRIPTION
**Pull request by andsel**

### CODE REVIEW

**Comment by jsvd** [comment]
@andsel can you rebase against master?

**Comment by jsvd** [comment]
I've tested this locally, and I can only see the pipeline id when running the ruby execution and not the default java execution, can you confirm that?

**Comment by andsel** [comment]
I've tested locally with java execution engine and I've seen the log line with the id. I've added some test log lines (then not committed) into the  java_pipeline.rb when the workers are fired up.
```
[2019-08-16T09:35:44,188][INFO ][logstash.javapipeline    ][another_test] Starting worker loop TODELETE
[2019-08-16T09:35:44,190][INFO ][logstash.javapipeline    ][another_test] Starting worker loop TODELETE
[2019-08-16T09:35:44,194][INFO ][logstash.javapipeline    ][another_test] Starting worker loop TODELETE
[2019-08-16T09:35:44,212][INFO ][logstash.javapipeline    ][another_test] Starting worker loop TODELETE
[2019-08-16T09:35:44,213][INFO ][logstash.javapipeline    ][test] Starting worker loop TODELETE
[2019-08-16T09:35:44,215][INFO ][logstash.javapipeline    ][test] Starting worker loop TODELETE
```

File: config/log4j2.properties [hunk 1]
```diff
@@ -4,7 +4,7 @@ name = LogstashPropertiesConfig
 appender.console.type = Console
 appender.console.name = plain_console
 appender.console.layout.type = PatternLayout
-appender.console.layout.pattern = [%d{ISO8601}][%-5p][%-25c] %m%n
+appender.console.layout.pattern = [%d{ISO8601}][%-5p][%-15c{1.}]%notEmpty{[%X{pipeline.id}]} %m%n
```
Comment (robbavey) [inline config/log4j2.properties]: Is the change to the logging category from `-25c` to `-15c{1.}` something we want to add, add just to console, or leave to another PR?

**Comment by andsel** [comment]
Fixes must be done also to another thread:
 https://github.com/elastic/logstash/blob/b8a5fee288cff186694c456e53dbcf9c563c8200/logstash-core/lib/logstash/java_pipeline.rb#L104-L115

**Comment by andsel** [comment]
jenkins test this please

File: logstash-core/lib/logstash/pipeline.rb [hunk 2]
```diff
@@ -338,6 +341,8 @@ def worker_loop(batch_size, batch_delay)
       end
       flush_filters_to_batch(batch, :final => false) if signal.flush?
       if batch.filteredSize > 0
+
+        java_import org.apache.logging.log4j.ThreadContext
```
Comment (colinsurprenant) [inline logstash-core/lib/logstash/pipeline.rb]: any reason to have this `java_import` here?

File: logstash-core/spec/logstash/pipeline_spec.rb [hunk 3]
```diff
@@ -15,6 +15,7 @@ def register
   end
 
   def run(queue)
+    @logger.debug("check log4j fish tagging: [" + ThreadContext.get("plugin.id") + "]")
```
Comment (colinsurprenant) [inline logstash-core/spec/logstash/pipeline_spec.rb]: are these debug lines in the spec still necessary or useful?

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java [hunk 4]
```diff
@@ -298,6 +303,22 @@ private static MethodLevelSyntaxElement callFilterFlush(final ClassFields fields
...
+        return () -> "org.apache.logging.log4j.ThreadContext.remove(\"plugin.id\")";
+    }
+
+    private static MethodLevelSyntaxElement setPluginIdForLog4j(final AbstractFilterDelegatorExt plugin) {
+        final ThreadContext context = RubyUtil.RUBY.getCurrentContext();
+        final IRubyObject configName = plugin.getConfigName(context);
+        return () -> "org.apache.logging.log4j.ThreadContext.put(\"plugin.id\", \"" + configName + "\")";
+    }
+
+    private static MethodLevelSyntaxElement setPluginIdForLog4j(final AbstractOutputDelegatorExt plugin) {
```
Comment (colinsurprenant) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java]: this code can probably be DRY'ed up for both method signatures?

**Comment by colinsurprenant** [comment]
Great work @andsel 👏  - left a few inline minor comments.

File: logstash-core/lib/logstash/pipeline.rb (hunk 2 above)
Comment (andsel) [reply logstash-core/lib/logstash/pipeline.rb]: no, was my typo, the import is already on top of the file

File: logstash-core/spec/logstash/pipeline_spec.rb (hunk 3 above)
Comment (andsel) [reply logstash-core/spec/logstash/pipeline_spec.rb]: yes it is need to prove that the ThreadContext contains the fish tag for the plugin, so inside the plugin I log the tag and then in the `expect` verify that the log contains the tag.

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java (hunk 4 above)
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java]: You are right, I try lower the repetition, the two methods are identical but one work on `AbstractFilterDelegatorExt` and the other on `AbstractOutputDelegatorExt` and the two classes hasn't got any common ancestor

File: config/log4j2.properties [hunk 5]
```diff
@@ -21,7 +21,7 @@ appender.rolling.policies.time.type = TimeBasedTriggeringPolicy
 appender.rolling.policies.time.interval = 1
 appender.rolling.policies.time.modulate = true
 appender.rolling.layout.type = PatternLayout
-appender.rolling.layout.pattern = [%d{ISO8601}][%-5p][%-25c] %-.10000m%n
+appender.rolling.layout.pattern = [%d{ISO8601}][%-5p][%-25c]%notEmpty{[%X{pipeline.id}]} %-.10000m%n
```
[the remark this answers is not in the document]
Comment (andsel) [reply config/log4j2.properties]: Added, thanks I missed to add to ConsoleAppender

File: config/jvm.options [hunk 6]
```diff
@@ -76,3 +76,6 @@
 
 # Entropy source for randomness
 -Djava.security.egd=file:/dev/urandom
+
+# Copy the logging context from parent threads to children
+-Dlog4j2.isThreadContextMapInheritable
```
[the remark this answers is not in the document]
Comment (andsel) [reply config/jvm.options]: You are right, fixed

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java [hunk 7]
```diff
@@ -196,12 +200,13 @@ private static Closure filterBody(final ValueSyntaxElement outputBuffer,
         final ValueSyntaxElement inputBuffer, final ClassFields fields,
         final AbstractFilterDelegatorExt plugin) {
         final ValueSyntaxElement filterField = fields.add(plugin);
-        final Closure body = Closure.wrap(
+        final Closure body = Closure.wrap(setPluginIdForLog4j(plugin),
```
[the remark this answers is not in the document]
Comment (andsel) [reply logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java]: Because method `filterDataset` invokes `filterBody` and filterBody sets and unsets the plugin.id in the context, while `outputDataset` create directly the statements creators, without invoking the filterBody. 
So in filterBody, I need to set to make working in filters while in outputDataset for output.

File: logstash-core/spec/logstash/pipeline_spec.rb (hunk 3 above)
Comment (andsel) [reply logstash-core/spec/logstash/pipeline_spec.rb]: Probably you are right, I can remove the "compiler way" that insert java code so set/unset the value in the ThreadContext, and move into the `doOutput` of `AbstractOutputDelegatorExt` (from `OutputDelegatorExt`), having a uniform way to set the plugin name for Ruby and Java plugins. What do you think?

**Comment by andsel** [comment]
@colinsurprenant @robbavey this PR going to split in 2, one for the `pipeline.id` and one for the `plugin.id`

**Comment by andsel** [comment]
I close this because it has been splitted in PR #11075 and PR #11078
